op-node/node/conductor.go

```diff
@@ -0,0 +1,73 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/metrics"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/driver"
+	"github.com/ethereum-optimism/optimism/op-service/dial"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/retry"
+	"github.com/ethereum/go-ethereum/log"
+
+	conductorRpc "github.com/ethereum-optimism/optimism/op-conductor/rpc"
+)
+
+// ConductorClient is a client for the op-conductor RPC service.
+type ConductorClient struct {
+	cfg       *Config
+	metrics   *metrics.Metrics
+	log       log.Logger
+	apiClient conductorRpc.API
+}
+
+// NoOpConductor is a no-op conductor that assumes this node is the leader sequencer.
+type NoOpConductor struct{}
+
+var _ driver.SequencerConductor = &NoOpConductor{}
+var _ driver.SequencerConductor = &ConductorClient{}
+
+// NewNoOpConductor returns a new conductor client for the op-conductor RPC service.
+func NewConductorClient(cfg *Config, log log.Logger, metrics *metrics.Metrics) *ConductorClient {
+	return &ConductorClient{cfg: cfg, metrics: metrics, log: log}
+}
+
+// Initialize initializes the conductor client.
+func (c *ConductorClient) Initialize() error {
```

Comment:
Is there a reason `NewConductorClient` doesn't do this initialise step? Typically an object returned from a constructor should be ready to use.

Reply:
This is mainly because sequencer & conductor need to talk to each other
1. conductor needs to know sequencer status / start / stop sequencer
2. sequencer need to talk to conductor to commit unsafe payload and check if it's leader

So one of them has to lazy initialize connection to the other during startup, and currently we chose to lazy startup in sequencer during `StartSequencer` call

Reply:
I've updated it to an alternative implementation where we explicitly check for leadership in the StartSequencer call, which lazily instantiates the apiClient internally. It will also ensure the client is re-used instead of creating a new one which should lessen the need for a Close call..

Alternatively, we could push the all of the conductor related code into the driver package and have the driver own the lifecycle of the conductor, lazily instantiate the ConductorClient which would have a typical client initialization in the constructor.

Reply:
We have this lazy initialization b/c the initialization can temporarily fail when reaching out to the network. It's a semi common pattern we use (though I do agree pushing failure up earlier would be nice)